lib/streamlit/commands/query_params.py

```diff
@@ -143,29 +143,102 @@ def _ensure_no_embed_params(
     return current_embed_params
 
 
+def _missing_key_error_message(key: str) -> str:
+    return f'st.query_params has no key "{key}". Did you forget to initialize it? '
+
+
 @dataclass
 class QueryParams:
     """A dict-like representation of query params.
-    The main difference is that it only stores and returns str and list[str].
+    The main difference is that it only stores and returns str and List[str].
 
     TODO(willhuang1997): Fill in these docs with examples and fix doc above. Above is just a stub for now.
     """
 
-    _query_params: Dict[str, List[Any]] = field(default_factory=dict)
+    _query_params: Dict[str, Union[List[str], str]] = field(default_factory=dict)
+
+    def __init__(self, query_params: Dict[str, Union[List[str], str]] = {}):
+        # avoid using ._query_params as that will use __setattr__
+        # and access _query_params which is None as _query_params isn't initialized
+        self.__dict__["_query_params"] = query_params
+
+    def _keys(self) -> Set[str]:
+        return set(self._query_params.keys())
+
+    def get(self, key: str, default: Any = None) -> Any:
+        if key in self._query_params:
+            return self._getitem(key)
+        else:
+            if default != None:
+                self._setitem(key, default)
+
+    def __iter__(self) -> Iterator[Any]:
+        return iter(self._keys())
 
-    def __init__(self, query_params: Dict[str, List[Any]] = {}):
-        self._query_params = query_params
+    def __getitem__(self, key: str) -> Any:
+        return self._getitem(key)
 
-    def __repr__(self):
-        return util.repr_(self)
+    def _getitem(self, key: str) -> str:
```

Comment:
Hm I think that's a good point; The reasoning as to why it was designed this way was because [we wanted it to be like the tornado `get_query_argument` ](https://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.get_query_argument)so I think we should maybe attempt to follow tornado imo off of first glance but I also didn't do much research into how people return query params tbh.

Reply:
The difference is that it doesn't feel as surprising to me to have that behavior with method calls like with tornado, but having the behavior in my code snippet with `=` and `==` feels absurd given natural expectations around how equality should work

CC @sfc-gh-jcarroll on this one since something just feels very wrong about assignment/equality comparison behaving like this under the current API

Reply:
We talked about this quite a bit in designing the API and at first pass I stand by the decision here. The assumption is that a developer is deciding whether a given param is a single value or a list and uses the appropriate get method accordingly. I see your point.

I think it's much more common to set a query param to a single value than a list. The current API has the absurd behavior you describe in that case.

```python
my_params = {"state": "California"}
st.experimental_set_query_params(my_params)
assert st.experimental_get_query_params() == my_params # fails because it returns {"state": ["California"]}
```

Reply:
I guess the alternate would be to not allow any dict-style get and only support methods calls on st.query_params. That feels heavy handed to me for the common case. I'm happy to talk further about it (maybe a discussion rather than async) if you think it merits revisiting the earlier product decisions.

Reply:
Hmm, I guess this may not be as bad as my gut feeling tells me since I'd imagine that the case where only a single query param is set should be overwhelmingly more common than the list case. I'll want to think about this more but don't think this is blocking despite my reservations